lxd/response/upgrade.go

```diff
@@ -0,0 +1,44 @@
+package response
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// Upgrade attempts to hijack w and send the HTTP 101 Swithing Protocols upgrade headers for protocolName.
+// Returns the upgraded connection once hijacked (even in some error scenarios), so if its returned it is the
+// reponsibility of the caller to close it.
+func Upgrade(w http.ResponseWriter, protocolName string) (net.Conn, error) {
+	hijacker, ok := w.(http.Hijacker)
+	if !ok {
+		return nil, fmt.Errorf("Webserver doesn't support hijacking")
+	}
+
+	conn, _, err := hijacker.Hijack()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to hijack connection: %w", err)
+	}
+
+	// Write the status line and upgrade header by hand since w.WriteHeader() would fail after Hijack().
+	sb := strings.Builder{}
+	sb.WriteString("HTTP/1.1 101 Switching Protocols\r\n")
+	sb.WriteString(fmt.Sprintf("Upgrade: %s\r\n", protocolName))
+	sb.WriteString("Connection: Upgrade\r\n\r\n")
+
+	conn.SetWriteDeadline(time.Now().Add(time.Second * 5))
+	n, err := conn.Write([]byte(sb.String()))
+	conn.SetWriteDeadline(time.Time{}) // Cancel deadline.
+
+	if err != nil {
+		return conn, fmt.Errorf("Failed writing upgrade headers: %w", err)
```

Comment:
`conn` will probably be `nil` here. Could we perhaps still write it as `nil`? Just for clarity.

Reply:
No it won't be nil as hijack succeded